server/graphql/v2/mutation/TransactionMutations.ts

```diff
@@ -2,16 +2,69 @@ import express from 'express';
 import { GraphQLNonNull, GraphQLString } from 'graphql';
 
 import orderStatus from '../../../constants/order_status';
-import { canReject } from '../../../graphql/common/transactions';
+import { TransactionKind } from '../../../constants/transaction-kind';
+import { FEES_ON_TOP_TRANSACTION_PROPERTIES } from '../../../constants/transactions';
 import { purgeCacheForCollective } from '../../../lib/cache';
 import { notifyAdminsOfCollective } from '../../../lib/notifications';
 import models from '../../../models';
-import { Forbidden, NotFound, Unauthorized } from '../../errors';
+import { canReject } from '../../common/transactions';
+import { Forbidden, NotFound, Unauthorized, ValidationFailed } from '../../errors';
 import { refundTransaction as legacyRefundTransaction } from '../../v1/mutations/orders';
+import { AmountInput } from '../input/AmountInput';
 import { fetchTransactionWithReference, TransactionReferenceInput } from '../input/TransactionReferenceInput';
 import { Transaction } from '../interface/Transaction';
 
 const transactionMutations = {
+  addPlatformTipToTransaction: {
+    type: new GraphQLNonNull(Transaction),
+    description: 'Add platform tips to a transaction',
+    args: {
+      transaction: {
+        type: new GraphQLNonNull(TransactionReferenceInput),
+        description: 'Reference to the transaction in the platform tip',
+      },
+      amount: {
+        type: GraphQLNonNull(AmountInput),
+        description: 'Amount of the platform tip',
+      },
+    },
+    async resolve(_: void, args, req: express.Request): Promise<typeof Transaction> {
+      if (!req.remoteUser) {
+        throw new Unauthorized('You need to be logged in to add a platform tip');
+      }
+
+      let transaction = await fetchTransactionWithReference(args.transaction, { throwIfMissing: true });
+
+      if (!req.remoteUser.isAdmin(transaction.HostCollectiveId)) {
+        throw new Unauthorized('Only host admins can add platform tips');
+      } else if (transaction.kind !== TransactionKind.ADDED_FUNDS) {
+        throw new ValidationFailed('Platform tips can only be added on added funds at the moment');
+      }
+
+      const isPlatformTipAvailable = await models.Transaction.findOne({
+        where: {
+          TransactionGroup: transaction.TransactionGroup,
+          kind: 'PLATFORM_TIP',
+        },
+      });
+
+      if (isPlatformTipAvailable) {
+        throw new Error('Platform tip is already set for this transaction group');
+      }
+
+      transaction = {
+        ...transaction,
+        data: { isFeesOnTop: true },
+        platformFeeInHostCurrency: args.amount.valueInCents,
+        FromCollectiveId: transaction.HostCollectiveId,
+        kind: TransactionKind.PLATFORM_TIP,
+        ...FEES_ON_TOP_TRANSACTION_PROPERTIES,
+      };
+
+      await models.Transaction.createFeesOnTopTransaction({ transaction, host: transaction.HostCollectiveId });
+      return transaction.dataValues;
```

Comment:
Why return the `dataValues` rather than the model? This will break if we try to fetch files like `fromAccount` on this transaction like in https://github.com/opencollective/opencollective-api/blob/69fb382095a65afe6d6da2554b5c2b7a83bed66a/server/graphql/v2/object/Debit.js#L17

Reply:
@Betree : I did this because otherwise trying to return the transaction I get the below error. Maybe because I am using the spread operator above to add some fields. But couldn't figure out exactly why; do you have any ideas? 🤔 

```
Error: Abstract type "Transaction" must resolve to an Object type at runtime for field "Mutation.addPlatformTipToTransaction" with value { dataValues: { id: 53096, type: "CREDIT", kind: "ADDED_FUNDS", uuid: "7f94aa79-46c0-4a1a-830a-22bf58185caa", description: "test", amount: 1800, currency: "USD", CreatedByUserId: 9476, FromCollectiveId: 2600, CollectiveId: 10886, HostCollectiveId: 10885, UsingGiftCardFromCollectiveId: null, OrderId: 6432, ExpenseId: null, hostCurrency: "USD", hostCurrencyFxRate: 1, amountInHostCurrency: 1800, platformFeeInHostCurrency: 0, hostFeeInHostCurrency: -72, paymentProcessorFeeInHostCurrency: 0, taxAmount: null, netAmountInCollectiveCurrency: 1728, data: [Object], TransactionGroup: "d1e9083e-4f22-4e6a-9b19-183f3a6d6b0d", RefundTransactionId: null, isRefund: false, createdAt: {}, updatedAt: {}, deletedAt: null, PaymentMethodId: 8768 }, _previousDataValues: { id: 53096, type: "CREDIT", kind: "ADDED_FUNDS", uuid: "7f94aa79-46c0-4a1a-830a-22bf58185caa", description: "test", amount: 1800, currency: "USD", CreatedByUserId: 9476, FromCollectiveId: 2600, CollectiveId: 10886, HostCollectiveId: 10885, UsingGiftCardFromCollectiveId: null, OrderId: 6432, ExpenseId: null, hostCurrency: "USD", hostCurrencyFxRate: 1, amountInHostCurrency: 1800, platformFeeInHostCurrency: 0, hostFeeInHostCurrency: -72, paymentProcessorFeeInHostCurrency: 0, taxAmount: null, netAmountInCollectiveCurrency: 1728, data: [Object], TransactionGroup: "d1e9083e-4f22-4e6a-9b19-183f3a6d6b0d", RefundTransactionId: null, isRefund: false, createdAt: {}, updatedAt: {}, deletedAt: null, PaymentMethodId: 8768 }, _changed: {}, _options: { isNewRecord: false, _schema: null, _schemaDelimiter: "", raw: true, attributes: [Array] }, isNewRecord: false, data: { isFeesOnTop: true }, platformFeeInHostCurrency: 0, FromCollectiveId: 10885, kind: "PLATFORM_TIP", CollectiveId: 8686, HostCollectiveId: 8686, hostCurrency: "USD", currency: "USD", paymentProcessorFeeInHostCurrency: NaN, amountInHostCurrency: NaN, amount: NaN }, received "undefined". Either the "Transaction" type should provide a "resolveType" function or each possible type should provide an "isTypeOf" function.
```

Reply:
@Betree : I've refactored the code at 695a02b to make sure we return the `Transaction` object. Let me know if this is good. 😄